src/Lib.P2P/AutoDialer.cs

```diff
@@ -0,0 +1,193 @@
+    /// </summary>
+    /// <remarks>
+    ///   Listens to the <see cref="SwarmService"/> and automically dials a
+    ///   new <see cref="Peer"/> when required.
+    /// </remarks>
+    public sealed class AutoDialer : IDisposable
+    {
+        private static readonly ILog Log = LogManager.GetLogger(typeof(AutoDialer));
+
+        /// <summary>
+        ///   The default minimum number of connections to maintain (16).
+        /// </summary>
+        public const int DefaultMinConnections = 16;
+
+        private readonly SwarmService _swarmService;
+        private int _pendingConnects;
+
+        /// <summary>
+        ///   Creates a new instance of the <see cref="AutoDialer"/> class.
+        /// </summary>
+        /// <param name="swarmService">
+        ///   Provides access to other peers.
+        /// </param>
+        public AutoDialer(SwarmService swarmService)
+        {
+            _swarmService = swarmService;
+            swarmService.PeerDiscovered += OnPeerDiscovered;
+            swarmService.PeerDisconnected += OnPeerDisconnected;
+        }
+
+        /// <summary>
+        ///  Releases the unmanaged and optionally managed resources.
+        /// </summary>
+        /// <param name="disposing">
+        ///   <b>true</b> to release both managed and unmanaged resources; <b>false</b> 
+        ///   to release only unmanaged resources.
+        /// </param>
+        private void Dispose(bool disposing)
+        {
+            if (!disposing)
+            {
+                return;
+            }
+            
+            _swarmService.PeerDiscovered -= OnPeerDiscovered;
+            _swarmService.PeerDisconnected -= OnPeerDisconnected;
+        }
+
+        /// <summary>
+        ///   Performs application-defined tasks associated with freeing, 
+        ///   releasing, or resetting unmanaged resources.
+        /// </summary>
+        public void Dispose() { Dispose(true); }
+
+        /// <summary>
+        ///   The low water mark for peer connections.
+        /// </summary>
+        /// <value>
+        ///   Defaults to <see cref="DefaultMinConnections"/>.
+        /// </value>
+        /// <remarks>
+        ///   Setting this to zero will basically disable the auto dial features.
+        /// </remarks>
+        public int MinConnections { get; set; } = DefaultMinConnections;
+
+#pragma warning disable VSTHRD100 // Avoid async void methods
+        /// <summary>
+        ///   Called when the swarm has a new peer.
+        /// </summary>
+        /// <param name="sender">
+        ///   The swarm of peers.
+        /// </param>
+        /// <param name="peer">
+        ///   The peer that was discovered.
+        /// </param>
+        /// <remarks>
+        ///   If the <see cref="MinConnections"/> is not reached, then the
+        ///   <paramref name="peer"/> is dialed.
+        /// </remarks>
+        private async void OnPeerDiscovered(object sender, Peer peer)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Remove this unused method parameter 'sender'.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)